File: 0437_path_sum_III/path_sum.java
public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

// Time: O(n)
// Space: O(n)
class Solution {
    public int pathSum(TreeNode root, int sum) {
        HashMap<Integer, Integer> cache = new HashMap();
        return preorder(root, 0, sum, cache);
    }

    public int preorder(TreeNode root, int currSum, int target, HashMap<Integer, Integer> cache) {
        if (root == null)
            return 0;
        currSum += root.val;
        int res = cache.getOrDefault(currSum - target, 0);
        if (currSum == target)
            res++;
        cache.put(currSum, cache.getOrDefault(currSum, 0) + 1);
        res += preorder(root.left, currSum, target, cache) + preorder(root.right, currSum, target, cache);
        cache.put(currSum, cache.get(currSum) - 1);
        return res;
    }
}

File: 0437_path_sum_III/path_sum.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def path_sum(root: TreeNode, target: int) -> int:
    """
    Time: O(n)
    Space: O(n)
    """
    return preorder(root, 0, target)


def preorder(node: TreeNode, curr_sum: int, target: int, cache=defaultdict(int)):
    if not node:
        return 0
    curr_sum += node.val
    res = cache[curr_sum - target]
    if curr_sum == target:
        res += 1
    cache[curr_sum] += 1
    res += preorder(node.left, curr_sum, target, cache) + \
        preorder(node.right, curr_sum, target, cache)
    cache[curr_sum] -= 1
    return res
